internal/xds/translator/ratelimit.go

```diff
@@ -0,0 +1,229 @@
+					},
+				}
+				action := &route.RateLimit_Action{
+					ActionSpecifier: &route.RateLimit_Action_HeaderValueMatch_{
+						HeaderValueMatch: &route.RateLimit_Action_HeaderValueMatch{
+							DescriptorValue: descriptorVal,
+							ExpectMatch: &wrapperspb.BoolValue{
+								Value: true,
+							},
+							Headers: []*route.HeaderMatcher{headerMatcher},
+						},
+					},
+				}
+				rlActions = append(rlActions, action)
+			}
+		}
+
+		// Case when header match is not set and the rate limit is applied
+		// to all traffic.
+		if len(rule.HeaderMatches) == 0 {
+			// Setup GenericKey action
+			action := &route.RateLimit_Action{
+				ActionSpecifier: &route.RateLimit_Action_GenericKey_{
+					GenericKey: &route.RateLimit_Action_GenericKey{
+						DescriptorKey:   getRateLimitDescriptorKey(descriptorPrefix, rIdx, -1),
+						DescriptorValue: getRateLimitDescriptorValue(descriptorPrefix, rIdx, -1),
+					},
+				},
+			}
+			rlActions = append(rlActions, action)
+		}
+
+		rateLimit := &route.RateLimit{Actions: rlActions}
+		rateLimits = append(rateLimits, rateLimit)
+	}
+
+	return rateLimits
+}
+
+func buildRateLimitServiceCluster(irListener *ir.HTTPListener) *cluster.Cluster {
+	// Return early if rate limits dont exist.
+	if !isRateLimitPresent(irListener) {
+		return nil
+	}
+
+	clusterName := getRateLimitServiceClusterName()
+	host, port := getRateLimitServiceGrpcHostPort()
+	rateLimitServerCluster := &cluster.Cluster{
+		Name:                 clusterName,
+		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_STRICT_DNS},
+		ConnectTimeout:       durationpb.New(10 * time.Second),
+		LbPolicy:             cluster.Cluster_RANDOM,
+		LoadAssignment: &endpoint.ClusterLoadAssignment{
+			ClusterName: clusterName,
+			Endpoints: []*endpoint.LocalityLbEndpoints{
+				{
+					LbEndpoints: []*endpoint.LbEndpoint{
+						{
+							HostIdentifier: &endpoint.LbEndpoint_Endpoint{
+								Endpoint: &endpoint.Endpoint{
+									Address: &core.Address{
+										Address: &core.Address_SocketAddress{
+											SocketAddress: &core.SocketAddress{
+												Address:       host,
+												PortSpecifier: &core.SocketAddress_PortValue{PortValue: uint32(port)},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		Http2ProtocolOptions: &core.Http2ProtocolOptions{},
+		DnsRefreshRate:       durationpb.New(30 * time.Second),
+		RespectDnsTtl:        true,
+		DnsLookupFamily:      cluster.Cluster_V4_ONLY,
+	}
+	return rateLimitServerCluster
+}
+
+func getRateLimitDescriptorKey(prefix string, ruleIndex, matchIndex int) string {
+	return prefix + "-key-rule-" + strconv.Itoa(ruleIndex) + "-match-" + strconv.Itoa(matchIndex)
+}
+
+func getRateLimitDescriptorValue(prefix string, ruleIndex, matchIndex int) string {
+	return prefix + "-value-rule-" + strconv.Itoa(ruleIndex) + "-match-" + strconv.Itoa(matchIndex)
+}
+
+func getRateLimitServiceClusterName() string {
+	return "ratelimit_cluster"
```

Comment:
is this a static cluster? if not, does EG allow end users use multiple RLSs?

Reply:
for the default RL feature, EG will build a static cluster, user can connect other RLS and build their own RL filters using https://github.com/envoyproxy/gateway/issues/24